lib/utils/report.cjs

```diff
@@ -126,6 +108,42 @@ function report(problem) {
 	result.warn(warningMessage, warningProperties);
 }
 
+/**
+ * @param {import('stylelint').Problem} problem
+ * @return {problem is import('stylelint').ProblemWithNormalizedPositions}
+ */
+function normalizeReportSourceLocation(problem) {
+	// Use the given node to calculate the start and end positions.
+	if (problem.node && (!problem.start || !problem.end)) {
+		const range = nodeRangeBy(problem.node, { index: problem.index, endIndex: problem.endIndex });
+
+		if (range) {
+			problem.start ??= range.start;
+			problem.end ??= range.end;
+		}
+	}
+
+	if (!validateTypes.isNumber(problem.line) && validateTypes.isNumber(problem.start?.line)) {
+		// Fallback the line to the line of the start position
+		problem.line = problem.start.line;
+	} else if (!problem.start && validateTypes.isNumber(problem.line)) {
+		// Fallback the start position to the line
+		problem.start = {
+			line: problem.line,
+			column: 1,
+		};
+	}
+
+	delete problem.index;
+	delete problem.endIndex;
+
+	if (problem.start && validateTypes.isNumber(problem.line)) {
```

Comment:
minor
question

If you pass a problem to `normalizeReportSourceLocation` without a `node` but with `start`, `end` and `line`.
This `if` will return `true` without throwing

> ```The "${ruleName}" rule failed to pass either a node or a line number to the \`report()\` function.```

correct?

Reply:
`node` is optional and we have test coverage for this.

see https://github.com/stylelint/stylelint/pull/8098#issuecomment-2453050157 for more details.